### PR TITLE
Predictions for regression in Datalab

### PR DESCRIPTION
**Pull request by elisno**
## Summary

Introduce the `predictions` argument to `RegressionLabelIssueManager.find_issues`.
The `pred_probs` argument will still be used on the Datalab-level. Internally, the `predictions` key will pick up the same value as the `pred_probs` key when setting up issue managers (for regression).

```python
# Example code snippet

from cleanlab import Datalab
import numpy as np
from sklearn.datasets import make_regression

# Generate some toy-examples
N = 100
X, y, W = make_regression(n_samples=N, n_features=3, noise=0.1, random_state=0, coef=True, bias=False)

# Get predictions from a regression model. Simplified for illustration.
y_pred = X @ W

# `predictions` used in find_issues
lab = Datalab(data={"y": y}, label_name='y', task="regression")
lab.find_issues(pred_probs=y_pred, issue_types= {"label": {}})

## Outputs
#Finding label issues ...
#
#Audit complete. 7 issues found in the dataset.
```

## Testing

> 🔍 Testing Done: Added specific tests for the RegressionLabelIssueManager. Write additional tests that run Datalab for regression with this new argument.

## Links to Relevant Issues or Conversations

- This is a follow-up to #796.
- This PR addresses #899.

```python
# How to set a threshold for flagging issues based on predictions
lab.find_issues(pred_probs=..., issue_types={"label":{"threshold": 1.0}}}
```

### CODE REVIEW

File: cleanlab/datalab/internal/issue_manager/regression/label.py [hunk 1]
```diff
@@ -38,6 +39,10 @@ class RegressionLabelIssueManager(IssueManager):
     clean_learning_kwargs :
         Keyword arguments to pass to the :py:meth:`CleanLearning <cleanlab.regression.learn.CleanLearning>` constructor.
 
+    threshold :
+        The threshold to use to determine if an example has a label issue. Only used if
```
Comment (jwmueller) [inline cleanlab/datalab/internal/issue_manager/regression/label.py:43]: this is a multiplier right not an absolute threshold

File: tests/datalab/test_datalab.py [hunk 2]
```diff
@@ -980,53 +980,109 @@ def test_pred_probs_precedence(self, pred_probs, random_embeddings):
...
+        X = regression_data["X"]
+        lab.find_issues(features=X)
+        lab.report()
+        summary = lab.get_issue_summary()
+
+        assert "label" in summary["issue_type"].values
+        assert (summary[summary["issue_type"] == "label"]["num_issues"] == 40).all()
+        assert np.isclose(
+            summary[summary["issue_type"] == "label"]["score"].values[0], 0.672974, atol=1e-5
+        )
```
Comment (jwmueller) [inline tests/datalab/test_datalab.py:1041]: can we assert which of the datapoints have issues, best test is assert that checks for high jaccard similarity between set of flagged issues and set of true issues

File: tests/datalab/test_datalab.py [hunk 3]
```diff
@@ -980,53 +980,109 @@ def test_pred_probs_precedence(self, pred_probs, random_embeddings):
...
+        model = KNeighborsRegressor(n_neighbors=5)
+        X = regression_data["X"]
+        lab.find_issues(
+            features=X, issue_types={"label": {"clean_learning_kwargs": {"model": model}}}
+        )
+        summary = lab.get_issue_summary()
+        assert (summary[summary["issue_type"] == "label"]["num_issues"] == 30).all()
+        assert np.isclose(
+            summary[summary["issue_type"] == "label"]["score"].values[0], 0.777305, atol=1e-5
+        )
```
Comment (jwmueller) [inline tests/datalab/test_datalab.py:1085]: can we assert which of the datapoints have issues, best test is assert that checks for high jaccard similarity between set of flagged issues and set of true issues

File: cleanlab/datalab/datalab.py [hunk 4]
```diff
@@ -188,6 +189,13 @@ def find_issues(
             To best detect label issues, provide this input obtained from the most accurate model you can produce.
 
             If provided, this must be a 2D array with shape (num_examples, K) where K is the number of classes in the dataset.
+            Must not be provided if `predictions` is provided.
```
Comment (jwmueller) [inline cleanlab/datalab/datalab.py:192]:
```suggestion
```

Comment (jwmueller) [reply cleanlab/datalab/datalab.py:192]: philosphy-wise, let's focus the current documentation mostly around `pred_probs` still being the primary input to Datalab. 

So keep the explanations of when to use `predictions` vs. pred_probs contained within the docstring of the `predictions` argument.

File: cleanlab/datalab/datalab.py [hunk 5]
```diff
@@ -188,6 +189,13 @@ def find_issues(
...
+
+        predictions :
+            Out-of-sample predicted output variables made by the model for every example in the dataset.
+            To best detect label issues, provide this input obtained from the most accurate model you can produce.
+
+            Must not be provided if `pred_probs` is provided.
```
Comment (jwmueller) [inline cleanlab/datalab/datalab.py:198]: needs more explanation about why this exists vs. just pred_probs and when to use one vs. the other

Comment (jwmueller) [reply cleanlab/datalab/datalab.py:198]: what is long-term plan here? 

deprecate `pred_probs` and have all tasks use `predictions`?

If so, I'd do followup PR where you allow classification tasks to work with `predictions` as well, internally you could just set:
```
if predictions is not None:
    pred_probs = predictions 
```

Comment (jwmueller) [reply cleanlab/datalab/datalab.py:198]: If ^ this is the long term plan, then will eventually need the `predictions` docstring to be quite involved for each task type. And should soon move the docstring for `pred_probs` into dosctring for `predictions` as well (in a subsection when the task is classification), so that `pred_probs` would start to be de-emphasized.

Alternatively if planning to keep `pred_probs` around for a while, want to think of plan on how to support both it and `predictions` in a nicer way

File: cleanlab/datalab/datalab.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (elisno) [reply cleanlab/datalab/datalab.py:192]: While resolving these things, I think it's better to look at pred_probs as the canonical/primary input for all tasks, unless specified otherwise (e.g. predictions for regression).

Comment (elisno) [reply cleanlab/datalab/datalab.py]: The safest approach here is to have a unified argument at the user-level. We'll keep `pred_probs` for all tasks, but don't have to modify the signatures of low-level classes like `RegressionLabelIssueManager`.

What these things have in common is that these are model-outputs that we use as inputs to Datalab. For  `pred_probs` and `predictions` in particular, they are just prediction-based model outputs.
I've added helper classes for handling each type of model output, but that won't change the top-level API for the user that use `Datalab.find_issues(pred_probs=...)`.
It's backwards compatible, easily extensible to other types of tasks and the other argument types like `features`, even `knn_graph`.

These helper classes are just an intermediate representation that won't get passed on to the issue managers directly, so this requires minimal changes.

Comment (elisno) [reply cleanlab/datalab/datalab.py]: This means I've removed the docstring for the `predictions` argument, but clarified the expected layout/structure of `pred_probs` in the relevant classes. That's much easier to maintain.

File: cleanlab/datalab/datalab.py (hunk 5 above)
Comment (elisno) [reply cleanlab/datalab/datalab.py]: We get the best of both worlds with the approach I briefly mention in https://github.com/cleanlab/cleanlab/pull/902#discussion_r1412323597.

We'll keep only `pred_probs` for backwards compatibility at the user-level.
The `RegressionLabelIssueManager` still uses the `predictions`  argument name, but these implementation details are handled in an intermediate stage when calling `Datalab.find_issues(pred_probs=...)`.
This gives us more time to think if we want to keep two, mutually exclusive arguments in `Datalab.find_issues` when one will suffice.

Docstring maintenance is also much simpler this way.

File: cleanlab/datalab/datalab.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (elisno) [reply cleanlab/datalab/datalab.py]: Removed.

File: tests/datalab/test_datalab.py (hunk 3 above)
Comment (jwmueller) [reply tests/datalab/test_datalab.py:1086]: this is not an end-to-end test though of what matters. What matters is that this code can accurately detect label issues. I'm not sure where that's being tested (maybe I overlooked)

File: cleanlab/datalab/datalab.py [hunk 6]
```diff
@@ -576,3 +588,34 @@ def load(path: str, data: Optional[Dataset] = None) -> "Datalab":
...
+            "regression": "predictions",
+        }
+        if task in task_tips:
+            error_msg += f" For a {task} task, consider passing `{task_tips[task]}` instead."
+    elif task == "classification" and predictions is not None:
+        error_msg = "`predictions` is not supported for classification tasks. Consider passing `pred_probs` instead."
+    elif task == "regression" and pred_probs is not None:
+        error_msg = "`pred_probs` is not supported for regression tasks. Consider passing `predictions` instead."
+    if error_msg:
+        raise ValueError(error_msg)
```
[the remark this answers is not in the document]
Comment (elisno) [reply cleanlab/datalab/datalab.py]: This piece of code has now been removed in favor of some new helper classes for validating the `pred_probs` inputs based on tasks.
Consider this resolved.

File: tests/datalab/test_datalab.py (hunk 2 above)
Comment (elisno) [reply tests/datalab/test_datalab.py:1042]: I've updated the dataset for testing and asserted which datapoints have issues. So this should be resolved.

The tricky thing is that the "real issues" might have smaller residuals and are not picked up (leading to a lower Jaccard similarity).

File: cleanlab/datalab/internal/issue_manager/regression/label.py (hunk 1 above)
Comment (elisno) [reply cleanlab/datalab/internal/issue_manager/regression/label.py]: Correct. The docstring has been updated.

File: tests/datalab/test_datalab.py (hunk 3 above)
Comment (elisno) [reply tests/datalab/test_datalab.py:1112]: I've updated the dataset for testing and asserted which datapoints have issues. So this should be resolved.

The tricky thing is that the "real issues" might have smaller residuals and are not picked up (leading to a lower Jaccard similarity).

**Review by huiwengoh — APPROVED** [review]
lgtm!

File: cleanlab/datalab/datalab.py [hunk 7]
```diff
@@ -187,7 +187,8 @@ def find_issues(
             Out-of-sample predicted class probabilities made by the model for every example in the dataset.
             To best detect label issues, provide this input obtained from the most accurate model you can produce.
 
-            If provided, this must be a 2D array with shape (num_examples, K) where K is the number of classes in the dataset.
+            If provided for classification, this must be a 2D array with shape (num_examples, K) where K is the number of classes in the dataset.
+            If provided for regression, this must be a 1D array with shape (num_examples,).
```
Comment (jwmueller) [inline cleanlab/datalab/datalab.py:191]:
```suggestion
            For classification data, this must be a 2D array with shape ``(num_examples, K)`` where ``K`` is the number of classes in the dataset.
            For regression data, this must be a 1D array with shape ``(num_examples,)`` containing the predicted value for each example.
```

File: docs/source/cleanlab/datalab/guide/custom_issue_manager.rst [hunk 8]
```diff
@@ -18,9 +18,10 @@ to detect a custom-defined type of issue alongside the pre-defined issue types i
         .. code-block:: python
 
             from cleanlab.datalab.internal.issue_manager_factory import register
-            register(MyIssueManager)
+            register(MyIssueManager)  # Defaults to task="classification"
+            # register(MyIssueManagerForRegression, task="regression") # For regression tasks
```
Comment (jwmueller) [inline docs/source/cleanlab/datalab/guide/custom_issue_manager.rst:22]:
```suggestion
            # register(MyIssueManagerForRegression, task="regression")  # Alternative for regression tasks
```

File: cleanlab/datalab/internal/issue_manager/regression/label.py [hunk 9]
```diff
@@ -58,50 +64,53 @@ def __init__(
         self,
         datalab: Datalab,
         clean_learning_kwargs: Optional[Dict[str, Any]] = None,
+        threshold: float = 0.1,
```
Comment (jwmueller) [inline cleanlab/datalab/internal/issue_manager/regression/label.py:67]:
```suggestion
        threshold: float = 0.13,
```

Comment (jwmueller) [reply cleanlab/datalab/internal/issue_manager/regression/label.py:67]: consider this alternate default value that would match near duplicates

Comment (jwmueller) [reply cleanlab/datalab/internal/issue_manager/regression/label.py:67]: disregard if that's a lot more work, or 0.1 is better default

File: cleanlab/datalab/internal/issue_manager/regression/label.py [hunk 10]
```diff
@@ -58,50 +64,53 @@ def __init__(
...
+        self.threshold = threshold
 
     def find_issues(
         self,
-        features: np.ndarray,
+        features: Optional[np.ndarray] = None,
+        predictions: Optional[np.ndarray] = None,
         **kwargs,
     ) -> None:
         """Find label issues in the datalab."""
```
Comment (jwmueller) [inline cleanlab/datalab/internal/issue_manager/regression/label.py:83]: you should write in english the priority order that happens inside here, eg:

If model is provided and features, then will use CleanLearning. Otherwise if pred_probs are provided, then will use them, etc etc. 

Otherwise this knowledge will be lost to time, if not explicitly documented

File: tests/datalab/test_datalab.py (hunk 2 above)
Comment (jwmueller) [reply tests/datalab/test_datalab.py:1042]: Sorry maybe I'm dense and don't see it. Where is the test that is checking for high jaccard similarity between set of flagged issues and set of true issues?

All I see is:

```
assert (summary[summary["issue_type"] == "label"]["num_issues"] == 40).all()
```

which is only testing how many issues were estimated (which we don't care about testing at all actually, we only care about testing which datapoints were flagged as issues and that the overlap between this set and the set of true erroneous values is sufficiently high).

File: tests/datalab/test_datalab.py [hunk 11]
```diff
@@ -981,53 +981,148 @@ def test_pred_probs_precedence(self, pred_probs, random_embeddings):
...
+    def regression_data(
+        self, num_examples=200, num_features=3, noise=0.01, error_frac=0.05, error_noise=10
+    ):
+        np.random.seed(SEED)
+        X = np.random.random(size=(num_examples, num_features))
+        coefficients = np.random.uniform(-1, 1, size=num_features)
+        label_noise = np.random.normal(scale=noise, size=num_examples)
+
+        true_y = np.dot(X, coefficients)
+        y = np.dot(X, coefficients) + label_noise
```
Comment (jwmueller) [inline tests/datalab/test_datalab.py:995]:
```suggestion
        true_y = np.dot(X, coefficients)
        y = np.dot(X, coefficients)
```

Comment (jwmueller) [reply tests/datalab/test_datalab.py:995]: why make our test more complicated by adding extra noise here?

File: tests/datalab/test_datalab.py [hunk 12]
```diff
@@ -981,53 +981,148 @@ def test_pred_probs_precedence(self, pred_probs, random_embeddings):
...
+
+        # add extra noisy examples
+        num_errors = int(num_examples * error_frac)
+        extra_noise = np.random.normal(scale=error_noise, size=num_errors)
```
Comment (jwmueller) [inline tests/datalab/test_datalab.py:999]:
```suggestion
        extra_noise = np.random.normal(scale=error_noise, size=num_errors) * max(1, error_noise)
```

Comment (jwmueller) [reply tests/datalab/test_datalab.py:999]: idea here is the test will be more stable if the extra_noise is more obvious. I don't think there's any reason to make this test challenging, it should be obvious errors IMO

**Review by jwmueller — COMMENTED** [review]
still had a couple of questions about the tests

File: cleanlab/datalab/internal/issue_manager/regression/label.py (hunk 10 above)
Comment (elisno) [reply cleanlab/datalab/internal/issue_manager/regression/label.py:83]: Added a note

image

File: tests/datalab/test_datalab.py (hunk 11 above)
Comment (elisno) [reply tests/datalab/test_datalab.py]: I've simplified the noise for this test, it seems to work now after some trial and error.

File: tests/datalab/test_datalab.py (hunk 2 above)
Comment (elisno) [reply tests/datalab/test_datalab.py]: The tests checking the high Jaccard similarity were placed behind those lines.

But now the summary assertions have been removed.
I've added some Jaccard Similarity assertions, as well as low FPR assertions as well.

File: tests/datalab/test_datalab.py (hunk 3 above)
Comment (elisno) [reply tests/datalab/test_datalab.py]: The tests checking the high Jaccard similarity were placed behind those lines.

But now the summary assertions have been removed.
I've added some Jaccard Similarity assertions, as well as low FPR assertions as well.

File: tests/datalab/test_datalab.py (hunk 12 above)
Comment (jwmueller) [reply tests/datalab/test_datalab.py]: in future tests like this, can also shrink size of `num_errors` so that it's not affecting the ML model too much during training

File: cleanlab/datalab/internal/issue_finder.py [hunk 13]
```diff
@@ -210,7 +211,8 @@ def find_issues(
             Out-of-sample predicted class probabilities made by the model for every example in the dataset.
             To best detect label issues, provide this input obtained from the most accurate model you can produce.
 
-            If provided, this must be a 2D array with shape (num_examples, K) where K is the number of classes in the dataset.
+            If provided for classification, this must be a 2D array with shape (num_examples, K) where K is the number of classes in the dataset.
+            If provided for regression, this must be a 1D array with shape (num_examples,).
```
Comment (jwmueller) [inline cleanlab/datalab/internal/issue_finder.py:215]:
```suggestion
            If provided for classification, this must be a 2D array with shape ``(num_examples, K)`` where K is the number of classes in the dataset.
            If provided for regression, this must be a 1D array with shape ``(num_examples,)``.
```

File: cleanlab/datalab/internal/issue_manager/regression/label.py (hunk 9 above)
Comment (jwmueller) [reply cleanlab/datalab/internal/issue_manager/regression/label.py:67]: @elisno did you try 0.13 instead of 0.1?

File: docs/source/cleanlab/datalab/internal/issue_manager/index.rst [hunk 14]
```diff
@@ -25,4 +25,11 @@ These are the issue managers that Datalab has not registered (yet).
 
 .. toctree::
     null
+
+Task-specific issue managers
+----------------------------
```
Comment (jwmueller) [inline docs/source/cleanlab/datalab/internal/issue_manager/index.rst:30]:
```suggestion
ML task-specific issue managers
---------------------------------
```

File: tests/datalab/test_datalab.py [hunk 15]
```diff
@@ -981,53 +981,144 @@ def test_pred_probs_precedence(self, pred_probs, random_embeddings):
...
+        # jaccard similarity
+        intersection = len(list(set(issue_ids).intersection(set(expected_issue_ids))))
+        union = len(set(issue_ids)) + len(set(expected_issue_ids)) - intersection
+        assert float(intersection) / union > 0.3
 
-    assert "label" in summary["issue_type"].values
-    assert (summary[summary["issue_type"] == "label"]["num_issues"] == 40).all()
+        # FPR
+        fpr = len(list(set(issue_ids).difference(set(expected_issue_ids)))) / len(issue_ids)
+        assert fpr < 0.3
```
Comment (jwmueller) [inline tests/datalab/test_datalab.py:1121]: open GH issue to improve this `fpr` in the future

File: cleanlab/datalab/internal/issue_manager/regression/label.py (hunk 9 above)
Comment (jwmueller) [reply cleanlab/datalab/internal/issue_manager/regression/label.py:67]: nevermind I see the false-positive rate is too high in the unit-test already, so let's not increase threshold

**Review by jwmueller — APPROVED** [review]
LGTM. Go ahead and merge after addressing my remaining minor comments

File: tests/datalab/test_datalab.py (hunk 15 above)
Comment (elisno) [reply tests/datalab/test_datalab.py:1121]: - [ ] Improve the stability of results from end-to-end tests of Datalab with label error-detection for regression tasks
